great_expectations/experimental/datasources/interfaces.py

```diff
@@ -371,94 +380,88 @@ def execution_engine_type(self) -> Type[ExecutionEngine]:
     #       should figure out a better way to anonymize and delete them.
-    _legacy_batch_markers: BatchMarkers
-    _legacy_batch_spec: BatchSpec
-    _legacy_batch_definition: BatchDefinition
+    batch_markers: "BatchMarkers" = Field(..., alias="legacy_batch_markers")
+    batch_spec: BatchSpec = Field(..., alias="legacy_batch_spec")
+    batch_definition: "BatchDefinition" = Field(..., alias="legacy_batch_definition")
 
-    def __init__(
-        self,
-        datasource: Datasource,
-        data_asset: DataAsset,
-        batch_request: BatchRequest,
-        # BatchDataType is Union[core.batch.BatchData, pd.DataFrame, SparkDataFrame].  core.batch.Batchdata is the
-        # implicit interface that Datasource implementers can use. We can make this explicit if needed.
-        data: BatchDataType,
-        # Legacy values that should be removed in the future.
-        legacy_batch_markers: BatchMarkers,
-        legacy_batch_spec: BatchSpec,
-        legacy_batch_definition: BatchDefinition,
-        # Optional arguments
-        metadata: Optional[Dict[str, Any]] = None,
-    ) -> None:
-        """This represents a batch of data.
-
-        This is usually not the data itself but a hook to the data on an external datastore such as
-        a spark or a sql database. An exception exists for pandas or any in-memory datastore.
-        """
-        # These properties are intended to be READ-ONLY
-        self._datasource: Datasource = datasource
-        self._data_asset: DataAsset = data_asset
-        self._batch_request: BatchRequest = batch_request
-        self._data: BatchDataType = data
-        self.metadata = metadata or {}
-
-        self._legacy_batch_markers = legacy_batch_markers
-        self._legacy_batch_spec = legacy_batch_spec
-        self._legacy_batch_definition = legacy_batch_definition
-
-        # computed property
-        # We need to unique identifier. This will likely change as I get more input
+    class Config:
+        allow_mutation = False
+        arbitrary_types_allowed = True
+        validate_assignment = True
+
+    @root_validator(pre=True)
+    def _set_id(cls, values: dict) -> dict:
+        # We need a unique identifier. This will likely change as we get more input.
         options_list = []
-        for k, v in batch_request.options.items():
+        for k, v in values["batch_request"].options.items():
             options_list.append(f"{k}_{v}")
+        values["id"] = "-".join(
+            [values["datasource"].name, values["data_asset"].name, *options_list]
+        )
+        return values
+
+    @overload
+    def head(
+        self, n_rows: Literal[0], fetch_all: Literal[False]
+    ) -> pd.DataFrame | list[pyspark_sql_Row]:
+        ...
+
+    @overload
+    def head(
+        self, n_rows: Optional[StrictInt] = None, fetch_all: StrictBool = False
+    ) -> pd.DataFrame | list[pyspark_sql_Row] | pyspark_sql_Row:
+        ...
```

Comment:
How do these overloads help us here?
The return types are the same.

Reply:
I believe if `n_rows == 0 and fetch_all = False` then the return type is `pd.DataFrame | list[pyspark_sql_Row]` otherwise it is `pd.DataFrame | list[pyspark_sql_Row] | pyspark_sql_Row`. Do we do any switching based on this or is it simpler to remove the `@overload`s?

Reply:
The return types are different if you squint 😄

Reply:
I had other overloads defined, but found them to be inaccurate due to some weird return type behavior with `pyspark` `head()`. I will remove this overload since it no longer seems to be adding much value.

Reply:
✅